substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/JavaUtilZipSubstitutions.java

```diff
@@ -187,89 +209,158 @@ private int deflateBytes(long addr, byte[] b, int off, int len, int flush) {
                 int res = ZLib.deflateParams(strm, level, strategy);
                 this.setParams = false;
                 if (res == ZLib.Z_OK()) {
-                    return Util_java_util_zip_Deflater_JDK8OrEarlier.update(this, len, strm);
+                    return Util_java_util_zip_Deflater_JDK10OrEarlier.update(this, len, strm);
                 } else if (res == ZLib.Z_BUF_ERROR()) {
                     return 0;
                 } else {
-                    throw new InternalError();
+                    throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
                 }
             } else {
                 int res = ZLib.deflate(strm, this.finish ? ZLib.Z_FINISH() : flush);
                 if (res == ZLib.Z_STREAM_END()) {
                     this.finished = true;
-                    return Util_java_util_zip_Deflater_JDK8OrEarlier.update(this, len, strm);
+                    return Util_java_util_zip_Deflater_JDK10OrEarlier.update(this, len, strm);
                 } else if (res == ZLib.Z_OK()) {
-                    return Util_java_util_zip_Deflater_JDK8OrEarlier.update(this, len, strm);
+                    return Util_java_util_zip_Deflater_JDK10OrEarlier.update(this, len, strm);
                 } else if (res == ZLib.Z_BUF_ERROR()) {
                     return 0;
                 } else {
-                    throw new InternalError();
+                    throw new InternalError(CTypeConversion.toJavaString(strm.msg()));
                 }
             }
         }
     }
 
+    // (JDK 11)src/java.base/share/native/libzip/Deflater.c-static jlong doDeflate(JNIEnv *env, jobject this, jlong addr,
+    //                       jbyte *input, jint inputLen,
+    //                       jbyte *output, jint outputLen,
+    //                       jint flush, jint params)
+    private static long doDeflate(long addr,
```

Comment:
(Now we are getting into things that are checked by the build process.)
This method needs an `@Substitute` annotation.

Reply:
Eclipse compiling the JDK-8 view of this code shows me 9 warnings.  I have not yet tried Eclipse with the JDK-11 view of this code.  Do I have to copy them to comments on this pull request, or can you reproduce (and fix \:-) them on your side?  The JDK-8 warnings are for field hidings, methods that could be declared static, unused parameters, and unnecessary casts.  Probably all those warnings should be suppressed unless you want to rename the names of locals from translated C functions so as not to hide the fields of the Java classes.

Reply:
I don't use Eclipse.  And, I'm not really interested in conforming to every potential reviewer's personal IDE settings in any event.  The main build seems to proceed with no warnings and no checkstyle problems, and my IDE (IntelliJ) is happy with the code.